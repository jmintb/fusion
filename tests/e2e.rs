use anyhow::Result;
use assert_cmd::Command;
use rstest::rstest;
use std::path::PathBuf;

#[rstest]
fn test_well_formed_programs(#[files("./test_programs/*.ts")] path: PathBuf) -> Result<()> {
    let mut cmd = Command::cargo_bin("typed_script")?;
    let assert = cmd.arg("run").arg(path.clone()).assert();
    assert.success();
    insta::assert_snapshot!(
        format!(
            "test_well_formed_programs_{}",
            path.file_name().unwrap().to_str().unwrap()
        ),
        String::from_utf8(cmd.ok().unwrap().stdout)?
    );
    Ok(())
}

#[rstest]
fn snapshop_mlir_output(#[files("./test_programs/*.ts")] path: PathBuf) -> Result<()> {
    let mut cmd = Command::cargo_bin("typed_script")?;
    let assert = cmd
        .args(&["build", "--emit-mlir"])
        .arg(path.clone())
        .assert();
    assert.success();
    insta::assert_snapshot!(
        format!(
            "test_mlir_snapshot_{}",
            path.file_name().unwrap().to_str().unwrap()
        ),
        String::from_utf8(cmd.ok().unwrap().stdout)?
    );
    Ok(())
}

#[rstest]
fn snapshop_ast_output(#[files("./test_programs/*.ts")] path: PathBuf) -> Result<()> {
    let mut cmd = Command::cargo_bin("typed_script")?;
    let assert = cmd
        .args(&["build", "--emit-ast"])
        .arg(path.clone())
        .assert();
    assert.success();
    insta::assert_snapshot!(
        format!(
            "test_ast_snapshot_{}",
            path.file_name().unwrap().to_str().unwrap()
        ),
        String::from_utf8(cmd.ok().unwrap().stdout)?
    );
    Ok(())
}
