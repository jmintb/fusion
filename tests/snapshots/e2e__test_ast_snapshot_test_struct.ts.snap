---
source: tests/e2e.rs
expression: "String::from_utf8(cmd.ok().unwrap().stdout)?"
---
[
    Decl(
        Function {
            keywords: [
                LlvmExtern,
            ],
            id: TSIdentifier(
                "fdopen",
            ),
            arguments: [
                FunctionArg {
                    name: TSIdentifier(
                        "fd",
                    ),
                    type: Integer,
                    access_mode: Let,
                },
                FunctionArg {
                    name: TSIdentifier(
                        "mode",
                    ),
                    type: String,
                    access_mode: Let,
                },
            ],
            body: None,
            return_type: Pointer,
        },
    ),
    Decl(
        Function {
            keywords: [
                LlvmExtern,
            ],
            id: TSIdentifier(
                "fclose",
            ),
            arguments: [
                FunctionArg {
                    name: TSIdentifier(
                        "fd",
                    ),
                    type: String,
                    access_mode: Let,
                },
            ],
            body: None,
            return_type: Unit,
        },
    ),
    Decl(
        Function {
            keywords: [
                LlvmExtern,
            ],
            id: TSIdentifier(
                "fwrite",
            ),
            arguments: [
                FunctionArg {
                    name: TSIdentifier(
                        "val",
                    ),
                    type: String,
                    access_mode: Let,
                },
                FunctionArg {
                    name: TSIdentifier(
                        "size",
                    ),
                    type: Integer,
                    access_mode: Let,
                },
                FunctionArg {
                    name: TSIdentifier(
                        "len",
                    ),
                    type: Integer,
                    access_mode: Let,
                },
                FunctionArg {
                    name: TSIdentifier(
                        "file",
                    ),
                    type: String,
                    access_mode: Let,
                },
            ],
            body: None,
            return_type: Integer,
        },
    ),
    Decl(
        Function {
            keywords: [
                LlvmExtern,
            ],
            id: TSIdentifier(
                "sprintf",
            ),
            arguments: [
                FunctionArg {
                    name: TSIdentifier(
                        "output",
                    ),
                    type: String,
                    access_mode: Let,
                },
                FunctionArg {
                    name: TSIdentifier(
                        "format",
                    ),
                    type: String,
                    access_mode: Let,
                },
                FunctionArg {
                    name: TSIdentifier(
                        "number",
                    ),
                    type: Integer,
                    access_mode: Let,
                },
            ],
            body: None,
            return_type: Integer,
        },
    ),
    Decl(
        Function {
            keywords: [
                LlvmExtern,
            ],
            id: TSIdentifier(
                "fflush",
            ),
            arguments: [
                FunctionArg {
                    name: TSIdentifier(
                        "file",
                    ),
                    type: String,
                    access_mode: Let,
                },
            ],
            body: None,
            return_type: Integer,
        },
    ),
    Decl(
        Function {
            keywords: [
                LlvmExtern,
            ],
            id: TSIdentifier(
                "sleep",
            ),
            arguments: [
                FunctionArg {
                    name: TSIdentifier(
                        "time",
                    ),
                    type: Integer,
                    access_mode: Let,
                },
            ],
            body: None,
            return_type: Integer,
        },
    ),
    Decl(
        Function {
            keywords: [
                Normal,
            ],
            id: TSIdentifier(
                "print",
            ),
            arguments: [
                FunctionArg {
                    name: TSIdentifier(
                        "val",
                    ),
                    type: String,
                    access_mode: Let,
                },
                FunctionArg {
                    name: TSIdentifier(
                        "len",
                    ),
                    type: Integer,
                    access_mode: Let,
                },
            ],
            body: Some(
                [
                    Assignment(
                        Assignment {
                            id: TSIdentifier(
                                "stdoutptr",
                            ),
                            expression: Call(
                                TSIdentifier(
                                    "fdopen",
                                ),
                                [
                                    Value(
                                        Integer(
                                            1,
                                        ),
                                        Integer,
                                    ),
                                    Value(
                                        String(
                                            "w",
                                        ),
                                        String,
                                    ),
                                ],
                            ),
                            typed_annotation: None,
                        },
                    ),
                    Expression(
                        Call(
                            TSIdentifier(
                                "fwrite",
                            ),
                            [
                                Value(
                                    Variable(
                                        TSIdentifier(
                                            "val",
                                        ),
                                    ),
                                    Unknown,
                                ),
                                Value(
                                    Variable(
                                        TSIdentifier(
                                            "len",
                                        ),
                                    ),
                                    Unknown,
                                ),
                                Value(
                                    Integer(
                                        1,
                                    ),
                                    Integer,
                                ),
                                Value(
                                    Variable(
                                        TSIdentifier(
                                            "stdoutptr",
                                        ),
                                    ),
                                    Unknown,
                                ),
                            ],
                        ),
                    ),
                    Expression(
                        Return(
                            Return {
                                expression: None,
                            },
                        ),
                    ),
                ],
            ),
            return_type: Unit,
        },
    ),
    Decl(
        Struct(
            StructType {
                identifier: TSIdentifier(
                    "Test",
                ),
                fields: [
                    StructField {
                        field_name: TSIdentifier(
                            "fielda",
                        ),
                        field_type: String,
                    },
                    StructField {
                        field_name: TSIdentifier(
                            "fieldb",
                        ),
                        field_type: String,
                    },
                    StructField {
                        field_name: TSIdentifier(
                            "fie",
                        ),
                        field_type: Integer,
                    },
                ],
            },
        ),
    ),
    Decl(
        Struct(
            StructType {
                identifier: TSIdentifier(
                    "String",
                ),
                fields: [
                    StructField {
                        field_name: TSIdentifier(
                            "val",
                        ),
                        field_type: Pointer,
                    },
                    StructField {
                        field_name: TSIdentifier(
                            "len",
                        ),
                        field_type: Integer,
                    },
                ],
            },
        ),
    ),
    Decl(
        Function {
            keywords: [
                Normal,
            ],
            id: TSIdentifier(
                "new_string",
            ),
            arguments: [
                FunctionArg {
                    name: TSIdentifier(
                        "val",
                    ),
                    type: String,
                    access_mode: Let,
                },
                FunctionArg {
                    name: TSIdentifier(
                        "len",
                    ),
                    type: Integer,
                    access_mode: Let,
                },
            ],
            body: Some(
                [
                    Expression(
                        Return(
                            Return {
                                expression: Some(
                                    Struct(
                                        TSIdentifier(
                                            "String",
                                        ),
                                        [
                                            Value(
                                                Variable(
                                                    TSIdentifier(
                                                        "val",
                                                    ),
                                                ),
                                                Unknown,
                                            ),
                                            Value(
                                                Variable(
                                                    TSIdentifier(
                                                        "len",
                                                    ),
                                                ),
                                                Unknown,
                                            ),
                                        ],
                                    ),
                                ),
                            },
                        ),
                    ),
                ],
            ),
            return_type: Named(
                TSIdentifier(
                    "String",
                ),
            ),
        },
    ),
    Decl(
        Function {
            keywords: [
                Normal,
            ],
            id: TSIdentifier(
                "print_string",
            ),
            arguments: [
                FunctionArg {
                    name: TSIdentifier(
                        "str",
                    ),
                    type: Named(
                        TSIdentifier(
                            "String",
                        ),
                    ),
                    access_mode: Let,
                },
            ],
            body: Some(
                [
                    Expression(
                        Call(
                            TSIdentifier(
                                "print",
                            ),
                            [
                                StructFieldRef(
                                    TSIdentifier(
                                        "str",
                                    ),
                                    TSIdentifier(
                                        "val",
                                    ),
                                ),
                                StructFieldRef(
                                    TSIdentifier(
                                        "str",
                                    ),
                                    TSIdentifier(
                                        "len",
                                    ),
                                ),
                            ],
                        ),
                    ),
                    Expression(
                        Return(
                            Return {
                                expression: None,
                            },
                        ),
                    ),
                ],
            ),
            return_type: Unit,
        },
    ),
    Decl(
        Function {
            keywords: [
                Normal,
            ],
            id: TSIdentifier(
                "main",
            ),
            arguments: [],
            body: Some(
                [
                    Assignment(
                        Assignment {
                            id: TSIdentifier(
                                "str",
                            ),
                            expression: Call(
                                TSIdentifier(
                                    "new_string",
                                ),
                                [
                                    Value(
                                        String(
                                            "ac \\n",
                                        ),
                                        String,
                                    ),
                                    Value(
                                        Integer(
                                            4,
                                        ),
                                        Integer,
                                    ),
                                ],
                            ),
                            typed_annotation: None,
                        },
                    ),
                    Expression(
                        Call(
                            TSIdentifier(
                                "print_string",
                            ),
                            [
                                Value(
                                    Variable(
                                        TSIdentifier(
                                            "str",
                                        ),
                                    ),
                                    Unknown,
                                ),
                            ],
                        ),
                    ),
                    Expression(
                        Return(
                            Return {
                                expression: None,
                            },
                        ),
                    ),
                ],
            ),
            return_type: Unit,
        },
    ),
]
